### PR TITLE
feat: visualize list of StateTransitionGraphs

### PR DESCRIPTION
**Pull request by redeboer**
Previously, DOT visualisation would crash if it got an uninitialised `StateTransitionGraph` (because it doesn't have edge properties). Now, the edge ID is used instead if there are no edge properties. In addition, if there are edge properties, the spin projection is printed as well. A `list` of solutions can now also be rendered in one graphviz file.

A quick demo is available under `test_graph.py`.

Also includes some minor fixes and refactorings to the `particle` submodule.

### CODE REVIEW

**Comment by redeboer** [comment]
Currently looking whether it's possible to render a list of `StateTransitionGraph` in *one* output file, but this does not have to be included in this PR.

File: expertsystem/state/particle.py [hunk 1]
```diff
@@ -526,27 +529,32 @@ def initialize_graph(
...
+            spin_projections = arange(  # type: ignore
+                -particle.state.spin, particle.state.spin + 1, 1.0
+            ).tolist()
+            if particle.mass == 0.0:
+                if 0.0 in spin_projections:
+                    del spin_projections[spin_projections.index(0.0)]
+            state = (particle_name, spin_projections)
+        return state
+
+    return [safe_set_spin_projections(state) for state in list_of_states]
```
Comment (spflueger) [inline expertsystem/state/particle.py:548]: Can you please avoid to do such unrelated refactoring work, if it is being overhauls by upcoming issues anyway

File: expertsystem/io/dot.py [hunk 2]
```diff
@@ -30,55 +32,88 @@ def write(instance: object, filename: str) -> None:
...
-        return ",".join(name_list)
 
+def __graph_to_dot(graph: StateTransitionGraph) -> str:
     dot_source = _DOT_HEAD
+    dot_source += __graph_to_dot_content(graph)
+    dot_source += _DOT_TAIL
+    return dot_source
 
+
+def __graph_to_dot_content(
```
Comment (spflueger) [inline expertsystem/io/dot.py:62]: Proposal: change this back to `__graph_to_dot`, but include the prefix part. Remove `__graph_to_dot`

File: expertsystem/io/dot.py [hunk 3]
```diff
@@ -30,55 +32,88 @@ def write(instance: object, filename: str) -> None:
...
 
-def __graph_to_dot(graph: StateTransitionGraph) -> str:
-    def node_name(edge_id: int, node_id: Optional[int] = None) -> str:
-        if node_id is None:
-            return f"edge{edge_id}"
-        return f"node{node_id}"
+def __graph_list_to_dot(graphs: List[StateTransitionGraph]) -> str:
+    dot_source = _DOT_HEAD
+    for i, graph in enumerate(graphs):
+        dot_source += __graph_to_dot_content(graph, prefix=f"g{i}_")
```
Comment (spflueger) [inline expertsystem/io/dot.py:50]: just call the proposed `__graph_to_dot` function

File: expertsystem/io/dot.py [hunk 4]
```diff
@@ -30,55 +32,88 @@ def write(instance: object, filename: str) -> None:
...
     outs = graph.get_final_state_edges()
-    for i in top:
-        dot_source += _DOT_DEFAULT_NODE.format(
-            node_name(i), graph.edge_props[i]["Name"]
-        )
-    for i in outs:
+    for edge_id in top + outs:
         dot_source += _DOT_DEFAULT_NODE.format(
-            node_name(i), graph.edge_props[i]["Name"]
+            prefix + __node_name(edge_id), __edge_label(graph, edge_id)
```
Comment (spflueger) [inline expertsystem/io/dot.py:70]: why are nodes labeled?

File: expertsystem/io/dot.py [hunk 5]
```diff
@@ -30,55 +32,88 @@ def write(instance: object, filename: str) -> None:
...
-    dot_source += _DOT_RANK_SAME.format(format_particle(outs))
-
+    dot_source += __rank_string(top, prefix)
+    dot_source += __rank_string(outs, prefix)
     for i, edge in graph.edges.items():
         j, k = edge.ending_node_id, edge.originating_node_id
         if j is None or k is None:
             dot_source += _DOT_DEFAULT_EDGE.format(
-                node_name(i, k), node_name(i, j)
+                prefix + __node_name(i, k), prefix + __node_name(i, j)
```
Comment (spflueger) [inline expertsystem/io/dot.py:78]: this is confusing. Im asking myself what a node_name has to do with edge labels

**Review by spflueger — COMMENTED** [review]
I proposed some changes to make it a bit cleaner. I'm not sure about this `node_name` stuff. Is a bit confusing to me.

File: expertsystem/io/dot.py (hunk 5 above)
Comment (redeboer) [reply expertsystem/io/dot.py:78]: This comes from #216

File: expertsystem/state/particle.py (hunk 1 above)
Comment (redeboer) [reply expertsystem/state/particle.py]: Kicked out

File: expertsystem/io/dot.py (hunk 2 above)
Comment (redeboer) [reply expertsystem/io/dot.py:62]: I have to think about this. Probably, the construction of wrapping the 'content' into `_DOT_HEAD` and a `_DOT_TAIL` can be better done with a decorator.

File: expertsystem/io/dot.py (hunk 3 above)
Comment (redeboer) [reply expertsystem/io/dot.py:65]: That's not how DOT syntax works

File: expertsystem/io/dot.py (hunk 2 above)
Comment (redeboer) [reply expertsystem/io/dot.py:75]: For the decorator, see 82ade244b2b8ab8031c870e936a5591dbc212faf
I'm not sure whether combining functions here makes much sense. How would you do it?

Comment (spflueger) [reply expertsystem/io/dot.py:74]: Well i meant, take this function rename it `__graph_to_dot` and also add the header and footer back in.

File: expertsystem/io/dot.py (hunk 3 above)
Comment (spflueger) [reply expertsystem/io/dot.py:65]: Ok, sorry I misunderstood the code then. So to clarify, the head and tail is only required for one single dot file, not for each graph?

File: expertsystem/io/dot.py (hunk 2 above)
Comment (spflueger) [reply expertsystem/io/dot.py:74]: I guess I misunderstood this DOT syntax then

File: expertsystem/io/dot.py (hunk 5 above)
Comment (spflueger) [reply expertsystem/io/dot.py:90]: yeah ok, but can we easily change that to something meaningful. Maybe I'm missing something here. What I'm saying is, cant we change this into `prefix + str(k), prefix + str(j)`?

Comment (spflueger) [reply expertsystem/io/dot.py:90]: Ah i think i get it now. You have to create some artificial nodes for the external lines. Ok, so lets keep it like this then and possibly refactor it in another PR.

File: expertsystem/io/dot.py (hunk 4 above)
Comment (redeboer) [reply expertsystem/io/dot.py:82]: The problem is that DOT language only works with nodes. So actually the node names are confusing: we prepend `"edge"` to a node that represents the end point of an open edge (initial and final state) and prepend `"node"` to internal nodes. See output of `test_dot.py`:

```dot
digraph {
    rankdir=LR;
    node [shape=point, width=0];
    edge [arrowhead=none];
    "edge0" [shape=none, label="J/psi(1S)[1]"];
    "edge2" [shape=none, label="gamma[1]"];
    "edge3" [shape=none, label="pi0[0]"];
    "edge4" [shape=none, label="pi0[0]"];
    { rank=same "edge0" };
    { rank=same "edge2", "edge3", "edge4" };
    "edge0" -> "node0";
    "node0" -> "node1" [label="f(0)(980)[0]"];
    "node0" -> "edge2";
    "node1" -> "edge3";
    "node1" -> "edge4";
}
```

File: expertsystem/io/dot.py (hunk 5 above)
Comment (redeboer) [reply expertsystem/io/dot.py:90]: It's a bit more complicated, but the labeling can indeed be improved https://github.com/ComPWA/expertsystem/pull/235#discussion_r478392499

**Review by spflueger — APPROVED** [review]
